src/logic/schedule-exporter/workers-absence-export.logic.ts

```diff
@@ -0,0 +1,162 @@
+
+export class WorkersAbsenceExportLogic {
+  private scheduleModel: ScheduleDataModel;
+
+  private revision: PrimaryMonthRevisionDataModel;
+
+  private cellsToMerge: number[][];
+
+  constructor(scheduleModel: ScheduleDataModel, revision: PrimaryMonthRevisionDataModel) {
+    this.scheduleModel = scheduleModel;
+    this.revision = revision;
+    this.cellsToMerge = [];
+  }
+
+  public setWorkersWorkSheet(workSheet: xlsx.Worksheet): void {
+    workSheet.pageSetup.showGridLines = true;
+    workSheet.pageSetup.fitToPage = true;
+    workSheet.pageSetup.fitToHeight = 1;
+    workSheet.pageSetup.fitToWidth = 1;
+    workSheet.pageSetup.horizontalCentered = true;
+
+    const workersInfoArray = this.createWorkersInfoSection(this.scheduleModel, this.revision);
+
+    const colLens = workersInfoArray[0].map((_, colIndex) =>
+      Math.max(...workersInfoArray.map((row) => row[colIndex]?.toString().length ?? 0))
+    );
+
+    workSheet.addRows(workersInfoArray);
+
+    this.cellsToMerge.forEach((pair, index) => {
+      if (this.cellsToMerge[index][0] > 0) {
+        const top = pair[0];
+        let bottom = pair[1];
+        let i = index;
+
+        while (this.cellsToMerge[i + 1] !== undefined && this.cellsToMerge[i + 1][0] === bottom) {
+          this.cellsToMerge[i + 1][0] = -1;
+          bottom = this.cellsToMerge[i + 1][1];
+          i++;
+        }
+
+        workSheet.mergeCells(top, 1, bottom, 1);
+        workSheet.mergeCells(top, 6, bottom, 6);
+        workSheet.mergeCells(top, 7, bottom, 7);
+      }
+    });
+
+    colLens.forEach((len, id) => {
+      workSheet.getColumn(id + 1).width = len + CELL_MARGIN;
+    });
+
+    workSheet.getColumn(1).alignment = { vertical: "middle", horizontal: "left" };
+    for (let i = 1; i <= ABSENCE_HEADERS.length; i++) {
+      workSheet.getColumn(i).alignment = { vertical: "middle", horizontal: "center" };
+    }
+
+    workSheet.getRow(1).alignment = { vertical: "middle", horizontal: "center" };
+    workSheet.getRow(1).font = { bold: true };
+  }
+
+  private createWorkersInfoSection(
+    scheduleModel: ScheduleDataModel,
+    revision: PrimaryMonthRevisionDataModel
+  ): (string | number)[][] {
+    const names = Object.keys(scheduleModel.employee_info?.type);
+    const month = scheduleModel.schedule_info.month_number;
+    const year = scheduleModel.schedule_info.year;
+    const dates = scheduleModel.month_info.dates;
+    const workers: (string | number)[][] = [];
+
+    const shiftTypes = scheduleModel.shift_types;
+
+    let employeeRowIndex = 2;
+
+    workers.push(ABSENCE_HEADERS);
+    workers.push(EMPTY_ROW);
+    names.forEach((name) => {
+      const workerShifts = scheduleModel.shifts[name] as ShiftCode[];
+      const workerContractType = scheduleModel.employee_info.contractType[name];
+      const { verboseDates } = new MonthInfoLogic(month, year, dates);
+
+      let moreThanOneRow = false;
+      workerShifts.forEach((shift, index) => {
```

Comment:
Dość długa i nieczytelna funkcja wyszła. Może by rozbić jakoś tę logikę na nowe funkcje. Łatwiej się będzie połapać o co chodzi. Z tego co przychodzi mi do głowy to do funkcji wynieść  np. 2 ścieżki z ifa, to co mamy w while'u